File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    rank = dict()
    
    for idx, player in enumerate(players):
        rank[player] = idx
    
    for player in callings:
        idx = rank[player]
        players[idx], players[idx - 1] = players[idx - 1], players[idx]
        rank[players[idx]] = idx
        rank[players[idx - 1]] = idx - 1
        
    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String, Integer> rankMap = new HashMap<>();
 
        for (int i = 0; i < players.length; i++) {
            rankMap.put(players[i], i);
        }
    
        for (String called : callings) {
            int rank = rankMap.get(called);
            
            String player = players[rank - 1];
            players[rank - 1] = players[rank];
            players[rank] = player;
            
            rankMap.put(players[rank - 1], rank - 1);
            rankMap.put(players[rank], rank);
        }
        
        return players;
    }
}